soccer/src/soccer/strategy/agent/position/position.hpp

```diff
@@ -114,6 +117,56 @@ class Position {
         communication::AgentPosRequestWrapper request);
     // END Agent-to-Agent Communication
 
+    /**
+     * @brief sends a direct pass request to the target robots
+     *
+     * @param target_robots a vector of robots to send a request to
+     */
+    void send_direct_pass_request(std::vector<u_int8_t> target_robots);
+
+    /**
+     * @brief receives and handles a pass_request
+     *
+     * @param pass_request the pass request to handle
+     * @return communication::PassResponse a response to the robot as to whether or not it
+     * is open for a pass
+     */
+    communication::PassResponse receive_pass_request(communication::PassRequest pass_request);
+
+    /**
+     * @brief tell another robot that this robot will pass to it
+     *
+     * @param target_robot the robot that will be passed to
+     */
+    void send_pass_confirmation(u_int8_t target_robot);
+
+    /**
+     * @brief acknowledges the pass confirmation from another robot
+     *
+     * @param incoming_pass_request the request that a ball will be coming to this robot
+     * @return communication::Acknowledge acknowledgement that the other robot may pass to this
+     * robot
+     */
+    virtual communication::Acknowledge acknowledge_pass(
+        communication::IncomingPassRequest incoming_pass_request);
+
+    /**
+     * @brief physically update the state of the robot to be passing the ball
+     *
+     * @param robot_id the robot id of the robot to pass the ball to
+     */
+    virtual void pass_ball(int robot_id);
```

Comment:
after reading the usage of this, I understand what you're going for here. however, currently you have it where the position implementations have to call the superclass implementation IN their own implementation of this virtual method, which is a little weird. you see this weirdness in being forced to add a Position::pass_ball() to force the superclass version.

instead, I would use the tmp-nvi template (see Position's get_task for more info). Position should have a concrete method (pass_ball()) that does all the shared messaging all derived classes need, then call a pure virtual function to force the positions to implement their own state update. something like:
```
void pass_ball(...) {
  // make pass request
  // make comm request
  derived_pass_ball(...);
}

virtual void pass_ball(...) = 0;
```
then in Defense/Off/Goalie:
```
void pass_ball() override ...
```